### PR TITLE
Added code to include WSL distros in profiles.

### PR DESCRIPTION
**Pull request by JBanks**
For #441 

Differences from #679 : 

- Variable names are now more descriptive;
- Excess test code has been removed;
- Code has been reorganized to make it easier to read;
- vcxproj has been updated to allow registry functions to be used.

### CODE REVIEW

**Comment by Biswa96** [comment]
Once @benhillis warned me not to use registry, it's not stable and documented.

**Comment by JBanks** [comment]
> Once benhillis warned me not to use registry, it's not stable and documented.

Yeah, we weren't too sure how else to get the information we needed.  Hopefully he can chime in and tell us of another way to go about this.

**Comment by DHowett-MSFT** [comment]
I chatted with @benhillis on this; we're engaging with our partners internally to figure out what the best course forward here is. :smile:

**Comment by DHowett-MSFT** [comment]
(Thanks for the contribution! We'll get to looking at it on Monday.)

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 1]
```diff
@@ -376,6 +425,34 @@ bool CascadiaSettings::_IsPowerShellCoreInstalled(std::wstring_view programFileE
...
+// Arguments:
+// - A string that contains the distributions directory.
+// - A string representing the distinct WSL distribution.
+// - A ref of a path that receives the result of [distribution].exe's full path.
+// Return Value:
+// - true or false.
+bool CascadiaSettings::_IsWSLDistributionAvailable(std::filesystem::path WSLDirectory, std::wstring searchString, std::filesystem::path& cmdline)
+{
+    // step through all of the directory names to find the one with our search string
+    for (auto& directory : std::filesystem::directory_iterator(WSLDirectory)) {
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: nit: I believe Terminal uses Allman-style braces; see #609.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 2]
```diff
@@ -376,6 +425,34 @@ bool CascadiaSettings::_IsPowerShellCoreInstalled(std::wstring_view programFileE
     return false;
 }
 
+// Function Description:
+// - Returns true if the executable for the distribution was found.
+// Arguments:
+// - A string that contains the distributions directory.
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: nit: distribution's

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 3]
```diff
@@ -151,6 +151,55 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
     
+    // Add all WSL distros from: HKCU\Software\Microsoft\Windows\CurrentVersion\Lxss
+    Profile WSLDistro{};
+    HKEY hLXSSKey;
+    // Should I use RegOpenCurrentUser to access the uncached profile?
+    // https://docs.microsoft.com/en-us/windows/desktop/api/winreg/nf-winreg-regopencurrentuser#remarks
+    if (RegOpenKeyExW(HKEY_CURRENT_USER, L"Software\\Microsoft\\Windows\\CurrentVersion\\Lxss",
+        0, KEY_READ, &hLXSSKey) == ERROR_SUCCESS)
+    {
+        std::filesystem::path AppsDirectory = ExpandEnvironmentVariableString(L"%ProgramFiles%");
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: Here and in other places where combined declaration and assignment in the new code is performed:
[ES.23: Prefer the {}-initializer syntax](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es23-prefer-the--initializer-syntax)

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 4]
```diff
@@ -151,6 +151,55 @@ void CascadiaSettings::_CreateDefaultProfiles()
     _profiles.emplace_back(powershellProfile);
     _profiles.emplace_back(cmdProfile);
     
+    // Add all WSL distros from: HKCU\Software\Microsoft\Windows\CurrentVersion\Lxss
+    Profile WSLDistro{};
+    HKEY hLXSSKey;
+    // Should I use RegOpenCurrentUser to access the uncached profile?
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: I can find no reason to not use RegOpenCurrentUser, but perhaps one of the console team members knows more?

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 5]
```diff
@@ -151,6 +151,55 @@ void CascadiaSettings::_CreateDefaultProfiles()
     _profiles.emplace_back(powershellProfile);
     _profiles.emplace_back(cmdProfile);
     
+    // Add all WSL distros from: HKCU\Software\Microsoft\Windows\CurrentVersion\Lxss
+    Profile WSLDistro{};
+    HKEY hLXSSKey;
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: To ensure the lifetime of this key handle is managed appropriately, be sure to use the [wil\resource.h helper types](https://github.com/microsoft/wil/blob/master/include/wil/resource.h):

wil::unique_hkey hLXSSKey;

This ensures that if execution exits this scope for any reason the resource is appropriately freed.  This also means that the manual call to RegCloseKey on 201 becomes unnecessary.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 6]
```diff
@@ -151,6 +151,55 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+        std::filesystem::path AppsDirectory = ExpandEnvironmentVariableString(L"%ProgramFiles%");
+        AppsDirectory /= "WindowsApps";
+        std::filesystem::path WSLCommandLine = L"";
+
+        wchar_t WSLKey[MAX_PATH] = L""; // Registry key for a distinct WSL Distribution
+        wchar_t DistributionName[MAX_PATH] = L""; // The name to be displayed by the flyout list
+        wchar_t PackageFamilyName[MAX_PATH] = L""; // The distinct name for the distribution
+
+        // These tell the Registry functions how much space they have to work with when returning a value.
+        DWORD WSLKey_Length = MAX_PATH;
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: In general, the Windows Terminal code uses the same style as the rest of Windows when it comes to naming variables.  In particular, I'd suggest a declaration and name that better fits the existing style would be something like:

DWORD cchWSLKey{std::size(WSLKey)};

The 'cch' here is special notation sometimes referred to as ["hungarian notation"](http://www.imm.dtu.dk/~alan/hungarian.html) used to indicate 'count of characters'.

Windows developers are used to relying on the name of the variable to tell them at a glance what the expected usage and related characteristics of a variable are in a concise fashion and so it would be helpful if the same could be applied to the new code here through this file.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 7]
```diff
@@ -151,6 +151,55 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+            RegEnumKeyExW(hLXSSKey, i, WSLKey, &WSLKey_Length, NULL, NULL, NULL, NULL) == ERROR_SUCCESS;
+            i++)
+        {
+            // Reset available space for each value.
+            WSLKey_Length = MAX_PATH;
+            DistributionName_Length = MAX_PATH;
+            PackageFamilyName_Length = MAX_PATH;
+
+            if (RegGetValueW(hLXSSKey, WSLKey, L"DistributionName", RRF_RT_REG_SZ, NULL,
+                DistributionName, &DistributionName_Length) == ERROR_SUCCESS
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: I don't know what the Terminal project's guidelines are around line length but splitting a multi-conditional across lines in this fashion may be challenging to read for some readers.  As such, consider either indenting the continuation or not wrapping the invocation to RegGetValueW across multiple lines.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 8]
```diff
@@ -151,6 +151,55 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+        wchar_t DistributionName[MAX_PATH] = L""; // The name to be displayed by the flyout list
+        wchar_t PackageFamilyName[MAX_PATH] = L""; // The distinct name for the distribution
+
+        // These tell the Registry functions how much space they have to work with when returning a value.
+        DWORD WSLKey_Length = MAX_PATH;
+        DWORD DistributionName_Length = MAX_PATH;
+        DWORD PackageFamilyName_Length = MAX_PATH;
+
+        for (int i = 0;
+            RegEnumKeyExW(hLXSSKey, i, WSLKey, &WSLKey_Length, NULL, NULL, NULL, NULL) == ERROR_SUCCESS;
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: Use nullptr instead where possible throughout this file, per this ISO C++ Guideline:
[ES.47: Use nullptr rather than 0 or NULL](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es47-use-nullptr-rather-than-0-or-null)

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 1 above)
Comment (binarycrusader) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Also, prefer const auto& where possible to clearly express intent [per ISO C++ guideline P3](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p3-express-intent).

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 9]
```diff
@@ -376,6 +425,34 @@ bool CascadiaSettings::_IsPowerShellCoreInstalled(std::wstring_view programFileE
...
+bool CascadiaSettings::_IsWSLDistributionAvailable(std::filesystem::path WSLDirectory, std::wstring searchString, std::filesystem::path& cmdline)
+{
+    // step through all of the directory names to find the one with our search string
+    for (auto& directory : std::filesystem::directory_iterator(WSLDirectory)) {
+        std::wstring FileString = directory.path().wstring();
+        if (FileString.find(searchString) != std::wstring::npos) {
+
+            // step through all of the files looking for an executable.  
+            // I couldn't find a way of determining the validity of this step.
+            for (auto& file : std::filesystem::directory_iterator(directory.path())) {
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: This seems plausible, but I would wait for the console team's feedback this coming week (hopefully) to see if there's a "supported" way of doing this.

**Comment by Biswa96** [comment]
I would suggest to first verify if this "WSL registry method" stable and make some docs so that other FOSS projects will be benefited from that. e.g. wsltty, LxRunOffline, ConEmu and many others.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 1 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Regarding const auto&, the PowerShell function above is also lacking it.  I think this should be included in a modernization or clarification PR, so I left it out of my changes.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 4 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: I updated it to include this function.  It won't be difficult to change it back.  I think it makes sense to use a live copy over a cached copy.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 2 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: I swear I've passed an English class before!

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 5 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Updated.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 6 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Updated.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 8 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Updated.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 3 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Thank you for taking the time to review my code and provide references for the corrections, it's been very educational.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 7 above)
Comment (binarycrusader) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: I've since been told they would like to keep line length around 100 characters.

**Comment by benhillis** [comment]
Thanks for taking the time to make this change, I think this will be a boon for the Windows Terminal.  That being said, I'd suggest not using the registry, but rather switching to using the output of "wsl.exe --list".  Here's what you should do at a high level.

1. Call CreateProcess on "%windir%\system32\wsl.exe --list", use a pipe as stdout.
2. Look at the process exit code, if it's non-zero stop (no distros).
3. Parse the output. Skip the first like (header), distros names cannot contain spaces, so read each line until the first space.

Here's a [snippet ](DistroLauncher/blob/540a593313f8abbc8ce3afe8ca98434e8a771798/DistroLauncher/DistributionInfo.cpp#L35)of code that should help get you started:
https://github.com/microsoft/WSL-

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 10]
```diff
@@ -151,6 +151,58 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+        DWORD cchWSLKey{ (DWORD) std::size(WSLKey) };
+        DWORD cchDistributionName{ (DWORD) std::size(DistributionName) };
+        DWORD cchPackageFamilyName{ (DWORD) std::size(PackageFamilyName) };
+
+        for (int i = 0;
+            RegEnumKeyExW(hLXSSKey.get(), i, WSLKey, &cchWSLKey, nullptr, nullptr, nullptr, nullptr) == ERROR_SUCCESS;
+            i++)
+        {
+            // Reset available space for each value.
+            cchWSLKey = (DWORD) std::size(WSLKey);
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: Try to use the least powerful cast possible to effect the change needed, so if possible, use C++-style casts here.  For example, `static_cast<DWORD>(std::size(WSLKey))`.  This should be changed here and in any new code added.

[ES.49: If you must use a cast, use a named cast](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es49-if-you-must-use-a-cast-use-a-named-cast)

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 11]
```diff
@@ -151,6 +151,58 @@ void CascadiaSettings::_CreateDefaultProfiles()
     _profiles.emplace_back(powershellProfile);
     _profiles.emplace_back(cmdProfile);
     
+    // Add all WSL distros from: HKCU\Software\Microsoft\Windows\CurrentVersion\Lxss
+    Profile WSLDistro{};
+    wil::unique_hkey hLXSSKey;
+    wil::unique_hkey hHKCUUncached;
+    RegOpenCurrentUser(KEY_READ, &hHKCUUncached); // Retrieve the uncached equivalent of HKEY_CURRENT_USER
```
Comment (binarycrusader) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: My reading of the msdn page for this function suggests that this is actually the *cached* result?

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 12]
```diff
@@ -151,6 +151,55 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+                && RegGetValueW(hLXSSKey, WSLKey, L"PackageFamilyName", RRF_RT_REG_SZ, NULL,
+                    PackageFamilyName, &PackageFamilyName_Length) == ERROR_SUCCESS) {
+                // The PackageFamilyName is split and has version information placed in the middle
+                // when creating the ProgramFiles directory.  I chose to only compare the last section.
+                std::wstring PFNString = PackageFamilyName;
+                std::wstring PFNSubString = PFNString.substr(PFNString.find(L"_"));
+                if (_IsWSLDistributionAvailable(AppsDirectory, PFNSubString, WSLCommandLine)) {
+                    WSLDistro.SetCommandline(WSLCommandLine);
+                    WSLDistro.SetName(DistributionName);
+                    WSLDistro.SetColorScheme({ L"Campbell" });
```
[the remark this answers is not in the document]
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: When I don't set the colour scheme I get the following in my profiles.json:

` "colorTable": [ "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000", "#000000" ],

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 11 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: > The HKEY_CURRENT_USER key maps to the root of the current user's branch in the HKEY_USERS key. It is cached for all threads in a process. Therefore, this value does not change when another user's profile is loaded.

The way I'm reading it is the `HKEY_CURRENT_USER `macro will use the path under `HKEY_USERS`.

> RegOpenCurrentUser uses the thread's token to access the appropriate key, or the default if the profile is not loaded.

I read this portion as saying that the `RegOpenCurrentUser `function has access to the user token allowing it to access the true/active `HKEY_CURRENT_USER`.

**Comment by JBanks** [comment]
> 
> 
> Thanks for taking the time to make this change, I think this will be a boon for the Windows Terminal. That being said, I'd suggest not using the registry, but rather switching to using the output of "wsl.exe --list". Here's what you should do at a high level.
> 
>     1. Call CreateProcess on "%windir%\system32\wsl.exe --list", use a pipe as stdout.
> 
>     2. Look at the process exit code, if it's non-zero stop (no distros).
> 
>     3. Parse the output. Skip the first like (header), distros names cannot contain spaces, so read each line until the first space.


@benhillis I'll see what I can get written up this week.  I can't seem to open those links you posted, but I'll try to find the information that you're referring to.

Assumptions that I'm going to make based on the comment from @YaSuenag https://github.com/microsoft/Terminal/issues/441#issuecomment-491295244

- Executable will be [name].exe
- Executable will be present in the user's path
- Executable will have punctuation marks removed (e.g. "Ubuntu-18.04" becomes `ubuntu1804.exe`)

Please let me know if anyone sees any problems or shortcomings with this.

**Comment by benhillis** [comment]
@jbanks - Sorry, fixed the [link](https://github.com/microsoft/WSL-DistroLauncher/blob/540a593313f8abbc8ce3afe8ca98434e8a771798/DistroLauncher/DistributionInfo.cpp#L35).

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 11 above)
Comment (ysc3839) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: https://docs.microsoft.com/en-us/windows/desktop/api/winreg/nf-winreg-regopenkeyexw
> If your service or application **impersonates different users**, do not use this function with HKEY_CURRENT_USER. Instead, call the RegOpenCurrentUser function.

I think RegOpenCurrentUser is for such condition, here we can use RegOpenKeyExW with HKEY_CURRENT_USER.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (mKay00) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: It probably gets initialized to #000000, if it doesn't exist. Like in https://github.com/microsoft/Terminal/pull/742#issuecomment-491870185

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 13]
```diff
@@ -213,6 +152,54 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+        HRESULT hr = CreateProcessW(nullptr, const_cast<LPWSTR>(command.c_str()), nullptr, nullptr,
+                                                        TRUE, CREATE_NO_WINDOW, nullptr, nullptr, &si, &pi);
+        if (SUCCEEDED(hr)) {
+            WaitForSingleObject(pi.hProcess, INFINITE);
+            DWORD exitCode;
+            if ((GetExitCodeProcess(pi.hProcess, &exitCode) == false) || (exitCode != 0)) {
+                hr = E_INVALIDARG;
+            }
+            if (SUCCEEDED(hr)) {
+                wchar_t buffer[500]; //arbitrary number.  Would like to evaluate other options for length of ouput
```
Comment (benhillis) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: Max distro name length is 256, plus another 10 for " (Default)", maybe round up to the next power of 2 to be safe (512).

Comment (Biswa96) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Where does the limit come from? How about using a defined value (e.g. UCHAR_MAX) instead of const value?

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 14]
```diff
@@ -213,6 +152,54 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+    if (CreatePipe(&readPipe, &writePipe, &sa, 0)) {
+        STARTUPINFO si{};
+        si.cb = sizeof(si);
+        si.dwFlags = STARTF_USESTDHANDLES;
+        si.hStdOutput = writePipe;
+        si.hStdError = writePipe;
+
+        PROCESS_INFORMATION pi{};
+        std::wstring command = L"wsl.exe --list";
+        HRESULT hr = CreateProcessW(nullptr, const_cast<LPWSTR>(command.c_str()), nullptr, nullptr,
```
Comment (benhillis) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: CreateProcess returns a bool, not an HRESULT.
https://msdn.microsoft.com/en-us/windows/desktop/ms682425

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 15]
```diff
@@ -273,19 +260,17 @@ void CascadiaSettings::_CreateDefaultKeybindings()
...
                               KeyChord{ KeyModifiers::Ctrl | KeyModifiers::Shift,
                                         static_cast<int>('9') });
+    keyBindings.SetKeyBinding(ShortcutAction::NewTabProfile9,
+                              KeyChord{ KeyModifiers::Ctrl | KeyModifiers::Shift,
+                                        static_cast<int>('0') });
 
     keyBindings.SetKeyBinding(ShortcutAction::ScrollUp,
                               KeyChord{ KeyModifiers::Ctrl | KeyModifiers::Shift,
-                                        VK_UP });
+                                        VK_PRIOR });
```
Comment (benhillis) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: This looks unrelated.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 16]
```diff
@@ -450,7 +438,7 @@ std::wstring CascadiaSettings::ExpandEnvironmentVariableString(std::wstring_view
     do
     {
         result.resize(requiredSize);
-        requiredSize = ::ExpandEnvironmentStringsW(source.data(), result.data(), gsl::narrow<DWORD>(result.size()));
+        requiredSize = ::ExpandEnvironmentStringsW(source.data(), result.data(), static_cast<DWORD>(result.size()));
```
Comment (benhillis) [inline src/cascadia/TerminalApp/CascadiaSettings.cpp]: Why was this changed?

Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Sorry, that's an artifact of my trying to learn git.  I didn't change this, something must have happened when I was bringing in the updated codebase between my revisions.

I think I need to scratch this branch and start anew, something went wrong when syncing this time and now I can't compile anymore.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 15 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: This is another artifact, sorry.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 17]
```diff
@@ -213,6 +152,54 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+        if (SUCCEEDED(hr)) {
+            WaitForSingleObject(pi.hProcess, INFINITE);
+            DWORD exitCode;
+            if ((GetExitCodeProcess(pi.hProcess, &exitCode) == false) || (exitCode != 0)) {
+                hr = E_INVALIDARG;
+            }
+            if (SUCCEEDED(hr)) {
+                wchar_t buffer[500]; //arbitrary number.  Would like to evaluate other options for length of ouput
+                DWORD bytesRead;
+                PeekNamedPipe(readPipe, buffer, (sizeof(buffer) - 1), &bytesRead, nullptr, nullptr);
```
[the remark this answers is not in the document]
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: This is something I meant to change, I had an issue with pipe inheritance and I was using that to test if there was any data in the pipe.  I'll change it to `readFile`.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 13 above)
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: @benhillis I'm using the buffer to grab the whole console output, so with my current code that could break some situations.  I'll see if I can figure out how to read a single line at a time from the pipe rather that pulling everything and then splitting the lines.

The other option I see is to have a huge buffer with `256 * MAX_DISTROS`, which is probably another fairly large number.

@Biswa96 `UCHAR_MAX `may be a little misleading (kind of like binarycrusader pointed out with my `NULL `vs `nullptr`).  Should I define a separate macro to define the maximum distribution name length for WSL distros?

Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: I've eliminated the buffer by figuring out how to convert the pipe into a stream using the following line:
`std::wifstream pipe{ _wfdopen( _open_osfhandle((intptr_t)readPipe, _O_WTEXT), L"r") };`

File: src/cascadia/TerminalApp/CascadiaSettings.cpp [hunk 18]
```diff
@@ -213,6 +152,54 @@ void CascadiaSettings::_CreateDefaultProfiles()
...
+                        std::getline(wlinestream, name, L' ');
+                        WSLDistro.SetCommandline(L"wsl.exe -d " + name);
+                        WSLDistro.SetColorScheme(L"Campbell");
+                        WSLDistro.SetName(name);
+                        _profiles.emplace_back(WSLDistro);
+                    }
+                }
+            }
+        }
+        CloseHandle(pi.hProcess);
```
[the remark this answers is not in the document]
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: I moved the 2 `pi` handles into the next level of if statements.

File: src/cascadia/TerminalApp/CascadiaSettings.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (JBanks) [reply src/cascadia/TerminalApp/CascadiaSettings.cpp]: Eliminated the use of the `HANDLE `to read the console output.
